File: frontend/src/types/riskQuestionnaire.ts
export type RiskQuestionnaireStatus = 'sent' | 'in_progress' | 'submitted';

export interface RiskQuestionnaireCapabilities {
    can_open: boolean;
    can_save_draft: boolean;
    can_submit: boolean;
    can_request_clarification: boolean;
    can_respond_to_clarifications: boolean;
}

export interface RiskQuestionnaireListItem {
    id: number;
    risk_id: number;
    risk_name?: string | null;
    assigned_to_user_id: number;
    sent_by_user_id: number;
    status: RiskQuestionnaireStatus;
    template_key: string;
    template_version: string;
    sent_at: string;
    due_at: string;
    submitted_at?: string | null;
    submitted_by_user_id?: number | null;

    assigned_to_user_name?: string | null;
    sent_by_user_name?: string | null;
    submitted_by_user_name?: string | null;
    capabilities?: RiskQuestionnaireCapabilities | null;
}

export interface RiskQuestionnaireDetail extends RiskQuestionnaireListItem {
    answers?: Record<string, unknown> | null;
    previous_submission?: RiskQuestionnairePreviousSubmission | null;
}

export interface RiskQuestionnairePreviousSubmission {
    id: number;
    submitted_at: string;
    template_version: string;
    answers?: Record<string, unknown> | null;
}

export interface RiskQuestionnaireDraftUpdate {
    answers: Record<string, unknown>;
}

export interface RiskQuestionnaireSubmit {
    answers: Record<string, unknown>;
}

export interface RiskQuestionnaireClarification {
    id: number;
    questionnaire_id: number;
    section_key: string;
    question_keys?: string[] | null;
    request_message: string;
    requested_by_user_id: number;
    requested_by_user_name?: string | null;
    requested_at: string;
    response_message?: string | null;
    responded_by_user_id?: number | null;
    responded_by_user_name?: string | null;
    responded_at?: string | null;
}

export interface RiskQuestionnaireClarificationCreate {
    section_key: string;
    request_message: string;
    question_keys?: string[] | null;
}

export interface RiskQuestionnaireClarificationRespond {
    response_message: string;
}

File: backend/app/schemas/risk_questionnaire.py
"""Pydantic schemas for risk questionnaires."""

from enum import Enum

from pydantic import BaseModel

from app.core.datetime_utils import UtcAwareDatetime


class RiskQuestionnaireStatusEnum(str, Enum):
    sent = "sent"
    in_progress = "in_progress"
    submitted = "submitted"


class RiskQuestionnaireCapabilitiesRead(BaseModel):
    can_open: bool = False
    can_save_draft: bool = False
    can_submit: bool = False
    can_request_clarification: bool = False
    can_respond_to_clarifications: bool = False


class RiskQuestionnaireListItemRead(BaseModel):
    id: int
    risk_id: int
    risk_name: str | None = None
    assigned_to_user_id: int
    sent_by_user_id: int
    status: RiskQuestionnaireStatusEnum
    template_key: str
    template_version: str
    sent_at: UtcAwareDatetime
    due_at: UtcAwareDatetime
    submitted_at: UtcAwareDatetime | None = None
    submitted_by_user_id: int | None = None

    assigned_to_user_name: str | None = None
    sent_by_user_name: str | None = None
    submitted_by_user_name: str | None = None
    capabilities: RiskQuestionnaireCapabilitiesRead | None = None

    model_config = {"from_attributes": True}


class RiskQuestionnairePreviousSubmissionRead(BaseModel):
    id: int
    submitted_at: UtcAwareDatetime
    template_version: str
    answers: dict[str, object] | None = None


class RiskQuestionnaireRead(RiskQuestionnaireListItemRead):
    answers: dict[str, object] | None = None
    previous_submission: RiskQuestionnairePreviousSubmissionRead | None = None


class RiskQuestionnaireDraftUpdate(BaseModel):
    answers: dict[str, object]


class RiskQuestionnaireSubmit(BaseModel):
    answers: dict[str, object]


class RiskQuestionnaireClarificationCreate(BaseModel):
    section_key: str
    request_message: str
    question_keys: list[str] | None = None


class RiskQuestionnaireClarificationRespond(BaseModel):
    response_message: str


class RiskQuestionnaireClarificationRead(BaseModel):
    id: int
    questionnaire_id: int
    section_key: str
    question_keys: list[str] | None = None

    request_message: str
    requested_by_user_id: int
    requested_by_user_name: str | None = None
    requested_at: UtcAwareDatetime

    response_message: str | None = None
    responded_by_user_id: int | None = None
    responded_by_user_name: str | None = None
    responded_at: UtcAwareDatetime | None = None

    model_config = {"from_attributes": True}
